File: backend/src/main/java/com/im/backend/service/FriendshipService.java
package com.im.backend.service;

import com.im.backend.model.Friendship;
import com.im.backend.model.User;

import java.util.List;

public interface FriendshipService {
    /**
     * 发送好友请求
     * @param requesterId 请求者ID
     * @param addresseeId 接收者ID
     * @return 创建的好友关系
     */
    Friendship sendFriendRequest(Long requesterId, Long addresseeId);
    
    /**
     * 接受好友请求
     * @param friendshipId 好友关系ID
     * @param userId 当前用户ID（必须是接收者）
     * @return 更新后的好友关系
     */
    Friendship acceptFriendRequest(Long friendshipId, Long userId);
    
    /**
     * 拒绝好友请求
     * @param friendshipId 好友关系ID
     * @param userId 当前用户ID（必须是接收者）
     * @return 更新后的好友关系
     */
    Friendship rejectFriendRequest(Long friendshipId, Long userId);
    
    /**
     * 获取用户的所有好友
     * @param userId 用户ID
     * @return 好友列表
     */
    List<User> getUserFriends(Long userId);
    
    /**
     * 获取用户收到的所有待处理好友请求
     * @param userId 用户ID
     * @return 待处理的好友请求列表
     */
    List<Friendship> getPendingFriendRequests(Long userId);
    
    /**
     * 检查两个用户是否是好友
     * @param userId1 用户1的ID
     * @param userId2 用户2的ID
     * @return 是否是好友
     */
    boolean areFriends(Long userId1, Long userId2);
    
    /**
     * 删除好友关系
     * @param userId 当前用户ID
     * @param friendId 要删除的好友ID
     * @return 是否删除成功
     */
    boolean deleteFriendship(Long userId, Long friendId);
}

File: mobile/lib/services/friendship_service.dart
import 'package:im_mobile/models/contact_model.dart';

import '../models/base_response.dart';
import '../models/friendship_model.dart';
import '../utils/http_client.dart';

class FriendshipService {
  static final HttpClient _httpClient = HttpClient();

  // 获取用户的所有好友
  static Future<List<ContactModel>> getUserFriends() async {
    BaseResponse<List<ContactModel>> response = await _httpClient.get<List<ContactModel>>(
        '/friendships/friends',
        fromJsonT: (json) => json == null 
          ? [] 
          : (json as List).map((item) => ContactModel.fromJson(item as Map<String, dynamic>)).toList(),
      );
      return response.data ?? [];
  }

  // 获取用户收到的所有待处理好友请求
 static Future<List<FriendshipModel>> getPendingFriendRequests() async {
    BaseResponse<List<FriendshipModel>> response = await _httpClient.get<List<FriendshipModel>>(
      '/friendships/pending',
      fromJsonT: (json) => json == null 
        ? [] 
        : (json as List).map((item) => FriendshipModel.fromJson(item as Map<String, dynamic>)).toList(),
    );
    return response.data ?? [];
  }

  // 发送好友请求
 static Future<BaseResponse<FriendshipModel?>> sendFriendRequest(int addresseeId) async {
    BaseResponse<FriendshipModel?> response = await _httpClient.post<FriendshipModel?>(
      '/friendships/request/$addresseeId',
      fromJsonT: (json) => json == null ? null : FriendshipModel.fromJson(json as Map<String, dynamic>),
    );
    return response;
  }

  // 接受好友请求
 static Future<FriendshipModel?> acceptFriendRequest(int friendshipId) async {
    BaseResponse<FriendshipModel> response = await _httpClient.put<FriendshipModel>(
      '/friendships/accept/$friendshipId',
      fromJsonT: (json) => FriendshipModel.fromJson(json as Map<String, dynamic>),
    );
    return response.data;
  }

  // 拒绝好友请求
static  Future<FriendshipModel?> rejectFriendRequest(int friendshipId) async {
    BaseResponse<FriendshipModel?> response = await _httpClient.put<FriendshipModel?>(
      '/friendships/reject/$friendshipId',
      fromJsonT: (json) => json == null ? null : FriendshipModel.fromJson(json as Map<String, dynamic>),
    );
    return response.data;
  }

  // 检查与指定用户是否是好友
 static Future<bool> checkFriendship(int userId) async {
    BaseResponse<bool> response = await _httpClient.get<bool>(
      '/friendships/check/$userId',
      fromJsonT: (json) => json == null ? false : json as bool,
    );
    return response.data ?? false;
  }

  // 删除好友关系
 static Future<bool> deleteFriendship(int friendId) async {
    BaseResponse<bool> response = await _httpClient.delete<bool>(
      '/friendships/$friendId',
      fromJsonT: (json) => json == null ? false : json as bool,
    );
    return response.data ?? false;
  }
}